### ISSUE TITLE
tf.nn.beam_search_decoder does not pick path with highest probability at next time step

### ISSUE THREAD

**Issue by sdrobert** [tensorflow/tensorflow#40727]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Colab Env (Linux Ubuntu 18.04)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
N/A
- TensorFlow installed from (source or binary):
Binary (I think)
- TensorFlow version (use command below):
v2.2.0-0-g2b96f3662b 2.2.0
- Python version:
3.6.9
- Bazel version (if compiling from source):
N/A
- GCC/Compiler version (if compiling from source):
N/A
- CUDA/cuDNN version:
N/A
- GPU model and memory:
None

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

The built-in CTC beam search decoder sometimes chooses a less probable path to keep in the beam by default than it could've when expanding a given step. In the Colab example provided below, I give a concrete example of when the path (0,) is retained in the beam instead of (0, 1, 0) despite the latter having a greater log-probability.

**Describe the expected behavior**

(0, 1, 0) should remain in the beam; (0,) should not.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

https://colab.research.google.com/drive/1i9gvj0VN2gMNloohbHW6ad3Ti7eiQQCM?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

When I was comparing against a simple python version [based off this Medium article](https://towardsdatascience.com/beam-search-decoding-in-ctc-trained-neural-networks-5a889a3d85a7), I noticed that whenever the results diverged, the root problem was always that a path that should've been kept in the beam wasn't. I suspect top-k sorting might be mucking up somewhere.

Thanks,
Sean

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/26897d798c5f763191014d5a558564e3/40727.ipynb). Thanks!

**Comment by sdrobert** [comment]
Hello,

While I can appreciate that you lot are very busy, it's been a month since I opened this issue and I'm afraid it's going to fall through the cracks. I would like to stress the importance of this issue.

I'm not sure why @amahendrakar removed the "bug" label and replaced it with "support." However, I would stress that (I believe) this is a mistake in the Tensorflow codebase, rather than something specific to my code, since i was able to reproduce it in a very  minimal gist. If I'm correct, the impact of this error is that all code using the built-in CTC beam search will output paths that are not as probable as they could be, potentially decreasing the overall accuracy reported in all sorts of research papers (including Google ones).

I originally stumbled on this when testing TF r1.15, which suggests that this is probably some logic error that has been around for some time in the code base. Either that, or I am fundamentally misunderstanding an aspect of this variety of search (AFAICT Graves did not publish a canon procedure for CTC beam search, leaving us to guess at appropriate rules).

Thank you for your time,
Sean

**Comment by yjiangling** [comment]
> Hello,
> 
> While I can appreciate that you lot are very busy, it's been a month since I opened this issue and I'm afraid it's going to fall through the cracks. I would like to stress the importance of this issue.
> 
> I'm not sure why @amahendrakar removed the "bug" label and replaced it with "support." However, I would stress that (I believe) this is a mistake in the Tensorflow codebase, rather than something specific to my code, since i was able to reproduce it in a very minimal gist. If I'm correct, the impact of this error is that all code using the built-in CTC beam search will output paths that are not as probable as they could be, potentially decreasing the overall accuracy reported in all sorts of research papers (including Google ones).
> 
> I originally stumbled on this when testing TF r1.15, which suggests that this is probably some logic error that has been around for some time in the code base. Either that, or I am fundamentally misunderstanding an aspect of this variety of search (AFAICT Graves did not publish a canon procedure for CTC beam search, leaving us to guess at appropriate rules).
> 
> Thank you for your time,
> Sean

Did you consider the prefix? It seems that the function will merge the probability if two beam have the same prfix, and finally obtain the highest merged probability path.

**Comment by sdrobert** [comment]
Hi @yjiangling,

I'm fairly certain that I haven't left out any probability mass because I'm relying on the op itself to calculate the probability of all but the pruned path. The way I calculate the pruned path, `[0, 1, 0]` at step t=3, given that `[0, 1]` and `[0]` are in the beam at step t=2, is to sum the total log-probability of the prefix `[0, 1]` (approx `-1.1840693`) with the log-probability of token `0` at t=3 (approx `0.16818666458129883`) to get approximately `-1.352256`. There's no need to separate the probability of `[0, 1]` into blank-ending and non-blank ending because `0` does not match the immediately preceding token `1`. This number is greater than the _reported_ probability of prefix `[0]` at t=3, `-1.8781054`. If there's a problem with the scores reported by Tensorflow before this point, well, I guess that'd be the bug :/.

By the way, this collab assumes the blank label is still `num_classes - 1` as per #42993. As soon as [this RTF](https://github.com/tensorflow/community/pull/295) merges, the collab will have to be revisited to get the indices right.

**Comment by yjiangling** [comment]
> Hi @yjiangling,
> 
> I'm fairly certain that I haven't left out any probability mass because I'm relying on the op itself to calculate the probability of all but the pruned path. The way I calculate the pruned path, `[0, 1, 0]` at step t=3, given that `[0, 1]` and `[0]` are in the beam at step t=2, is to sum the total log-probability of the prefix `[0, 1]` (approx `-1.1840693`) with the log-probability of token `0` at t=3 (approx `-0.16818666458129883`) to get approximately `-1.352256`. There's no need to separate the probability of `[0, 1]` into blank-ending and non-blank ending because `0` does not match the immediately preceding token `1`. This number is greater than the _reported_ probability of prefix `[0]` at t=3, `-1.8781054`. If there's a problem with the scores reported by Tensorflow before this point, well, I guess that'd be the bug :/.
> 
> By the way, this collab assumes the blank label is still `num_classes - 1` as per #42993. As soon as [this RTF](https://github.com/tensorflow/community/pull/295) merges, the collab will have to be revisited to get the indices right.

Yes，the CTC Loss in tensorflow 2.X the indices 0 is set to be blank as default，but the decode function greedy_decode and beam_search_decode is still set num_classes-1 as blank and have no option to set it.

**Comment by foxik** [comment]
I think there really needs to be a bug in the decoder. If you run the original Colab with beam size `K=3`, the `[0, 1, 0]` is returned with logprob `-1.352256`, which is larger than the logprob of `[0]` at timestep `t=3`; and the `[0, 1, 0]` cannot use the additional element `[1]` in the beam at timestep `t=2`.

Would it be possible to update the labels to include `bug` and also include `TF-2.4`?

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by sdrobert** [comment]
Hi Venkat,

As you can see, I made this bug report 4 years ago, and received no response from @ebrevdo. While I understand that the devs are under no obligation to fix or even acknowledge open bugs, asking me to keep this issue up-to-date for an empty audience is offensive, to say the least.

All the best,
Sean
